### PR TITLE
PYTHON-3366 Support mypy 0.971 and test with latest version

### PR DESCRIPTION
**Pull request by ShaneHarvey**
https://jira.mongodb.org/browse/PYTHON-3366 

Changes:
- Support mypy 0.971 and test with latest version.
- Fix:
   ```
    bson/__init__.py: note: In member "decode" of class "BSON":
    bson/__init__.py:1311: error: A function returning TypeVar should receive at
	least one argument containing the same Typevar  [type-var]
	    ...ptions: CodecOptions = DEFAULT_CODEC_OPTIONS) -> _DocumentType:  # typ...
   ```

### CODE REVIEW

File: bson/__init__.py [hunk 1]
```diff
@@ -1308,7 +1308,7 @@ def encode(
         """
         return cls(encode(document, check_keys, codec_options))
 
-    def decode(self, codec_options: CodecOptions = DEFAULT_CODEC_OPTIONS) -> _DocumentType:  # type: ignore[override]
+    def decode(self, codec_options: "CodecOptions[_DocumentType]" = DEFAULT_CODEC_OPTIONS) -> _DocumentType:  # type: ignore[override,assignment]
```
Comment (ShaneHarvey) [inline bson/__init__.py:1311]: The "assignment" ignore is required to silence this error: 
```
bson/__init__.py: note: In member "decode" of class "BSON":
bson/__init__.py:1311: error: Incompatible default for argument "codec_options" (default has type "CodecOptions[MutableMapping[str, Any]]", argument
has type "CodecOptions[_DocumentType]")  [assignment]
        def decode(self, codec_options: "CodecOptions[_DocumentType]" = DEFAULT_CODEC_OPTIONS) -> _DocumentType:  # type: ignore[override]
```

File: pymongo/pyopenssl_context.py [hunk 2]
```diff
@@ -192,7 +192,7 @@ def __init__(self, protocol):
         # side configuration and wrap_socket tries to support both client and
         # server side sockets.
         self._callback_data.check_ocsp_endpoint = True
-        self._ctx.set_ocsp_client_callback(callback=_ocsp_callback, data=self._callback_data)
+        self._ctx.set_ocsp_client_callback(callback=_ocsp_callback, data=self._callback_data)  # type: ignore[arg-type]
```
Comment (ShaneHarvey) [inline pymongo/pyopenssl_context.py:195]: This ignore is required until https://github.com/python/typeshed/pull/8371 is fixed and released.

**Comment by ShaneHarvey** [comment]
The remaining errors are in the tests and look like:
```
test/test_bson.py: note: In member "assertInvalid" of class "TestBSON":
test/test_bson.py:120: error: Argument 2 to "assertRaises" of "TestCase" has incompatible type
"Callable[[Union[bytes, memoryview, mmap, array[Any]], Optional[CodecOptions[_DocumentType]]], _DocumentType]"; expected "Callable[..., object]" 
[arg-type]
            self.assertRaises(InvalidBSON, decode, data)
                                           ^
```

Looks like a mypy bug. I'm not sure why it considers these types are incompatible.

**Review by blink1073 — APPROVED** [review]
LGTM!

**Pull request merged**